### PR TITLE
Replication cache for sharing hydrated messages

### PR DESCRIPTION
**Pull request by vytautas-karpavicius**
**What changed?**
Added replication cache to store hydrated replication messages.

Two components were introduced.
- `replication.Cache` - Data structure only. This is in memory implementation of a cache. Replication tasks always come in order, thus internally heap is utilized to achieve faster `Ack` (range removals from cache). Overall complexity:
  - Get: O(1)
  - Put: O(log N)
  - Ack: O(log N)
where N is the cache size. Cache capacity is configure via new dynamic config property `history.replicatorCacheCapacity`. For now default is 10K. This may be changed later. Increasing capacity can be done dynamically, reducing - requires a restart to reclaim memory.
- `replication.TaskStore` - "manager" component for the cache. It manages separate cache for each cluster. Putting new message will put it to all caches which accept it. Ack/Get work with individual caches. It also deals with filtering out irrelevant task for domains not participating in that cluster. Rate limiting logic and retries were also moved here out from `TaskAckManager`. 

**Why?**
This is an optimization attempt to reduce replication caused load on the database.

In particular every time we fetch replication messages, they go through hydration process where raw replication tasks are enriched with history event data (at least for `ReplicationTaskTypeHistoryV2` messages, which are the majority). This results in `readhistorybranch` persistence calls.

When we have multiple polling clusters in nDC setup, each of them result is fetching and hydrating same replication messages. This can be optimized, by introducing this cache. Once one cluster hydrates the message its pointer is put to caches for other clusters, so that they do not have to repeat this process.

We can even take a step further. After a successful transaction for creating/updating workflow, we already have persisted history blobs at that point. Thus on success we could cheaply hydrate the messages and put them into cache immediately. #4953 - preparation step for that. Another PR will follow after, to complete the integration.

On a healthy system this should significantly reduce `readhistorybranch` persistence calls. If replication on one cluster starts lagging, the cache will fill up and we will stop accepting new messages for it. That means that later it will need to read them from the database as it currently done today.

**How did you test it?**
- 100% unit test coverage for new components: `replication.Cache` and `replication.TaskStore`
- Passing existing integration tests
- Deployed branch on staging2, to verify replication still works and cache metrics are emitted.

**Potential risks**

**Release notes**

**Documentation Changes**

### CODE REVIEW

**Comment by davidporter-id-au** [comment]
Stupid question: Why would these tasks be read multiple times during replication? 

Excellent description by the way

File: service/history/replication/task_store.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+		}
+
+		scope := m.scope.Tagged(metrics.SourceClusterTag(cluster))
+
+		err := cache.Put(task)
+		switch err {
+		case errCacheFull:
+			scope.IncCounter(metrics.CacheFullCounter)
+		case errAlreadyAcked:
+			// No action, this is expected.
```
Comment (Shaddoll) [inline service/history/replication/task_store.go:187]: Is this branch needed? How could a new history event be acked by a remote cluster?

**Review by Shaddoll — APPROVED** [review]

File: service/history/replication/cache.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+// Cache is an in-memory implementation of a cache for storing hydrated replication messages.
+// Messages can come out of order as long as their task ID is higher than last acknowledged message.
+// Out of order is expected as different source clusters will share hydrated replication messages.
+//
+// Cache utilizes heap to keep replication messages in order. This is needed for efficient acknowledgements in O(log N).
+//
+// Cache capacity can be increased dynamically. Decrease will require a restart, as new tasks will not be accepted, but memory will not be reclaimed either.
+//
+// Cache methods are thread safe. It is expected to have writers and readers from different go routines.
+type Cache struct {
```
Comment (Shaddoll) [inline service/history/replication/cache.go:48]: Looks like it's worth making it an interface, considering we probably will introduce disk cache later.
But it can be done later, not necessarily in this PR

**Comment by vytautas-karpavicius** [comment]
> Stupid question: Why would these tasks be read multiple times during replication?
> 
> Excellent description by the way

In nDC scenario, when we have >1 polling cluster, each is reading replication tasks and hydrating them. Once hydrated it is cached so that others could reuse it.

File: service/history/replication/task_store.go (hunk 1 above)
Comment (vytautas-karpavicius) [reply service/history/replication/task_store.go:187]: It could be removed, but I kept it to explicitly cover all possible errors returned by the cache.

Suppose we have current cluster X and two polling clusters A and B.  Both have cache size 3. Reads are in batch size 2.
1. X commit tasks 1, 2.
2. A reads until 2, hydrates them and puts to cache A and cache B. size(A) = 2, size(B) = 2
3. A acks until 2, size(A) = 0, ack(A) = 2
4. X commit task 3, 4.
5. A reads until 4, hydrates them and puts to cache A and cache B. size(A) = 2, size(B) = 3. Task4 is skipped with errCacheFull in cache B.
6. A acks until 4. size(A) = 0, ack(A) = 4
7. B reads until 2, returns from cache. size(B) = 3
8. B acks until 2, size(B) = 1, ack(B) = 2
9. B reads until 4, returns Task3 from cache, hydrates Task4 and puts to cache A and cache B. Cache A skips it with errAlreadyAcked, since ack(A) = 4 > Task3. size(B) = 2
10. B acks until 4, size(B) = 0, ack(B) = 4

To sum up. Some polling cluster may be too slow and gets it cache full. After hydrating caches for faster polling clusters will reject them, since they were already acked by them. Let me know if you need further clarification.

File: service/history/replication/cache.go (hunk 2 above)
Comment (vytautas-karpavicius) [reply service/history/replication/cache.go:48]: Generally in GoLang (as opposed to some other languages) we should avoid introducing interfaces until we actually need them. It is also a good practice to define them in consumer package instead (unless they are very common interfaces, used in multiple places).

https://github.com/golang/go/wiki/CodeReviewComments#interfaces
https://bryanftan.medium.com/accept-interfaces-return-structs-in-go-d4cab29a301b

File: service/history/replication/task_store.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+			// Some cluster(s) may be already past this, due to different fetch rates.
+		}
+
+		scope.UpdateGauge(metrics.CacheSize, float64(cache.Size()))
```
Comment (mantas-sidlauskas) [inline service/history/replication/task_store.go:191]: there's always caveat with gauges. Let's say we have 10 history instances and all of those will report gauge metric with different value. Usually, last one wins, but it will be not meaningful at all.

Add additional, "shard" tag to expose cache size per shard

File: service/history/replication/cache_test.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+	require.NoError(t, cache.Put(task4))
+	assert.Equal(t, 1, cache.Size())
+	assert.Nil(t, cache.Get(100))
+	assert.Nil(t, cache.Get(200))
+	assert.Nil(t, cache.Get(300))
+	assert.Equal(t, task4, cache.Get(400))
+}
+
+func BenchmarkCache(b *testing.B) {
```
Comment (davidporter-id-au) [inline service/history/replication/cache_test.go:102]: nice!

File: service/history/replication/task_ack_manager.go [hunk 5]
```diff
@@ -121,25 +96,8 @@ func (t *TaskAckManager) GetTasks(ctx context.Context, pollingCluster string, la
...
-		// construct replication task from DB
-		_ = t.rateLimiter.Wait(ctx)
-		var replicationTask *types.ReplicationTask
-		op := func() error {
-			var err error
-			replicationTask, err = t.hydrator.Hydrate(ctx, *task)
-			return err
-		}
-		err = t.throttleRetry.Do(ctx, op)
 		switch err.(type) {
```
Comment (davidporter-id-au) [inline service/history/replication/task_ack_manager.go:101]: consider using `errors.Is` here

File: service/history/replication/task_store.go [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/common/quotas"
+	"github.com/uber/cadence/common/types"
+	"github.com/uber/cadence/service/history/config"
+)
+
+var errUnknownCluster = errors.New("unknown cluster")
```
Comment (davidporter-id-au) [inline service/history/replication/task_store.go:43]: Consider exporting errors like this so we can match on them elsewhere without doing string matching

File: service/history/replication/task_store.go [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+	}
+
+	domain, err := m.domains.GetDomainByID(info.DomainID)
+	if err != nil {
+		return nil, fmt.Errorf("resolving domain: %w", err)
+	}
+
+	// Domain does not exist in this cluster, do not replicate the task
+	if !domain.HasReplicationCluster(cluster) {
+		return nil, nil
```
Comment (davidporter-id-au) [inline service/history/replication/task_store.go:123]: Is this an expected scenario ever?

File: service/history/replication/task_store.go (hunk 3 above)
Comment (davidporter-id-au) [reply service/history/replication/task_store.go:191]: Could m3 tolerate that kind of cardinality? I'm not sure it could. But your point about this being therefore not helpful if it's last-write-wins is a good one

**Review by davidporter-id-au — APPROVED** [review]

File: service/history/replication/task_store.go (hunk 7 above)
Comment (vytautas-karpavicius) [reply service/history/replication/task_store.go:123]: It is. Domain can be configured on only on a subset of clusters. No need to replicate them everywhere. Although this is probably not very common.

File: service/history/replication/task_store.go (hunk 3 above)
Comment (vytautas-karpavicius) [reply service/history/replication/task_store.go:191]: It can, but barely. We already have a few places that add shard tag. In particular DLQ size, and replication backlog, fetched tasks, etc. in replication/task_ack_manager.go.

File: service/history/replication/cache.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+	"sync"
+
+	"github.com/uber/cadence/common/dynamicconfig"
+	"github.com/uber/cadence/common/types"
+)
+
+var (
+	errCacheFull    = errors.New("cache is full")
+	errAlreadyAcked = errors.New("already acked")
+)
```
Comment (Groxx) [inline service/history/replication/cache.go:37]: probably worth adding "replication cache: ..." to these, since the message itself is pretty generic / could conceivably be from other cache / ack code.  it'd take fairly rich error wrapping to differentiate them, currently, and we don't yet have _any_ noteworthy wrapping.
